### PR TITLE
Add GET /servers API

### PR DESCRIPTION
**Pull request by mriedem**
This adds a `GET /servers` admin API which attempts to be as close to the `GET /users` admin API with just the servers resource portion of that response without needing to query the users table in the database.

A difference from the servers returned in the `GET /users` API is that `GET /servers` will return the `id` and minimal `user` object for each server resource in the response. The `id` is needed to uniquely identify a server in upcoming `GET /servers/{id}` and `DELETE /servers/{id}` APIs without the user context. The user name is just a nicety to use other APIs like `DELETE /users/{name}/servers/{server_name}`.

This is eventually something that is going to be used when culling idle servers without querying users to improve performance in environments with a large number of users (tens of thousands).

Part of issue #2954

### CODE REVIEW

**Comment by mriedem** [comment]
The new `test_get_servers` test is failing because the tests are not idempotent so by the time it runs there are other servers (spawners) in the database. I guess the servers tests should be moved before the users tests?

File: docs/rest-api.yml [hunk 1]
```diff
@@ -768,6 +778,9 @@ definitions:
   Server:
     type: object
     properties:
+      id:
```
Comment (mriedem) [inline docs/rest-api.yml]: Question for reviewers: how are API changes like this communicated? Change log? Release notes? IOW, how does someone know that they can or should see these fields, i.e. maybe I don't see them because JupyterHub I'm talking to is old, or maybe because I'm not an admin and so I can't do `GET /servers`?

File: docs/rest-api.yml [hunk 2]
```diff
@@ -768,6 +778,9 @@ definitions:
...
+        type: integer
+        description: The unique internal ID of the server.
```
Comment (mriedem) [inline docs/rest-api.yml]: I suppose this should also say:

```
Only available on ``GET /servers``.
```

File: jupyterhub/apihandlers/base.py [hunk 3]
```diff
@@ -131,9 +132,16 @@ def write_error(self, status_code, **kwargs):
...
+            # TODO(mriedem): Would be nice to join the user field from
+            # the Spawner ORM object when we fetch it from the DB so we don't
+            # need the roundtrip.
+            orm_user = orm.User.find_by_id(self.db, spawner.user_id)
+            user = self._user_from_orm(orm_user)
+            spawner = spawner_obj.Spawner(orm_spawner=spawner, user=user)
+        server = {
             'name': spawner.name,
             'last_activity': isoformat(spawner.orm_spawner.last_activity),
             'started': isoformat(spawner.orm_spawner.started),
```
Comment (mriedem) [inline jupyterhub/apihandlers/base.py]: Note that with the `Spawner.__getattr__` override now we don't need to touch the `orm_spawner` directly here. Future cleanup I guess.

File: jupyterhub/apihandlers/servers.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+class ServerListAPIHandler(APIHandler):
+    @admin_only
+    def get(self):
+        # TODO(mriedem): Should include_state be conditional on
+        # `spawner.active` like how `user_model` works for
+        # `GET /users`?
+        include_state = True
+        data = [
+            self.server_model(
+                spawner, include_state=include_state, include_user_and_id=True
```
Comment (mriedem) [inline jupyterhub/apihandlers/servers.py]: I sort of wonder if state and the user-related fields like `url` and `user` should be query parameters which are off by default so that someone calling `GET /servers` that don't care about any of the user information don't need to fetch it from the DB. IOW, I tried to make `GET /servers` similar to how `GET /users` works but maybe that should be opt-in and we should be returning the most minimal representation of each server/spawner by default?

File: pytest.ini [hunk 5]
```diff
@@ -10,3 +10,4 @@ markers =
     services: mark as a services test
     user: mark as a test for a user
     slow: mark a test as slow
+    server: mark as a test for a server
```
Comment (mriedem) [inline pytest.ini]: Note that I'm not sure how useful this is because while it allows running just the server annotated tests, because of how the tests are setup to not be idempotent just running `pytest -m server` will fail, so this isn't very useful right now but seemed like the pattern to copy.

File: docs/rest-api.yml (hunk 1 above)
Comment (mriedem) [reply docs/rest-api.yml]: Should I be updating `docs/source/changelog.md`?

Comment (mriedem) [reply docs/rest-api.yml]: Thinking through this a bit more, the root `GET /` route returns the API version so if API changes are documented in the change log properly then I guess people could determine what they should expect that way in client side code though it would be nice if the API reference itself indicated when something was added or changed by version.

As for knowing if `GET /servers` is available that should be pretty easy - if it's not the response will be a `404` so the client needs to fallback to some other behavior.

File: jupyterhub/apihandlers/servers.py (hunk 4 above)
Comment (mriedem) [reply jupyterhub/apihandlers/servers.py]: While adding the `DELETE /servers/{id}` API in pull #2967 I realized [1] that it's not going to work well in the current form of the API because you need the server id to delete it but only `GET /servers` in this pull request returns the server id and `GET /servers` is an admin-only API, so this means non-admins can't delete their own servers unlike `DELETE /users/{name}/servers/{server_name}` which is bad UX.

So I think our options are:

1. Continue with adding a dedicated `DELETE /servers/{id}` API which allows `@admin_or_self` and modify the `GET /user` and `GET /users/{name}` responses so they include the id of the user-owned servers.

    * Pros: admin tooling like `cull_idle_servers` would not need to get the user information when listing servers and deleting them making `GET /servers` faster (no extra join from `servers` to `users` in the DB).
    * Cons: more API changes and bloat (more than one way to delete a server), expanded maintenance cost

2. Scrap the idea of a dedicated `DELETE /servers/{id}` API and as long as `GET /servers` returns the user name the client can use the existing `DELETE /users/{name}/servers/{server_name}` API.

    * Pros: re-use existing API to delete a server that `cull_idle_servers` is already using, no new `DELETE /servers/{id}` API or changes to `GET /user` or `GET /users/{name}` to return the server id.
    * Cons: Slight performance hit on `GET /servers` to include the user name since we have to query that from the DB per server that is returned (joining the user when we query the servers might negate that, or we could cache the `user_id -> User` mapping if the same user owns more than one server).

In the case of issue #2954 the client is the `cull_idle_servers` script which runs with an admin token.

[1] https://github.com/jupyterhub/jupyterhub/pull/2967/files#diff-0a701bf780442df25926c4771a5f170eR30

File: jupyterhub/orm.py [hunk 6]
```diff
@@ -220,6 +220,15 @@ def find(cls, db, name):
         """
         return db.query(cls).filter(cls.name == name).first()
 
+    @classmethod
+    def find_by_id(cls, db, id):
```
Comment (mriedem) [inline jupyterhub/orm.py]: Note that we won't need this if we join the user with the servers when we fetch the servers from the DB. Maybe that's already done though with this?

```python
_orm_spawners = relationship(
    "Spawner", backref="user", cascade="all, delete-orphan"
)
```

File: jupyterhub/apihandlers/servers.py (hunk 4 above)
Comment (mriedem) [reply jupyterhub/apihandlers/servers.py]: I've gone with option 2 and dropped pull #2967 once I got `cull_idle_servers` updated to use `GET /servers` in pull #2969.

Comment (rkdarst) [reply jupyterhub/apihandlers/servers.py]: At least I use `state` in my hubs to adjust culling parameters within the cull_idle script.  It could be reasonable to be off by default, as long as it's obvious how to turn it on again, which I can add later if need be (but I guess it should be the same setting for /users and /servers?).

Comment (mriedem) [reply jupyterhub/apihandlers/servers.py]: > At least I use `state` in my hubs to adjust culling parameters within the cull_idle script. It could be reasonable to be off by default, as long as it's obvious how to turn it on again, which I can add later if need be

I would think that would be a query parameter if `state` is not returned by default from `GET /servers` so you could do something like:

`GET /servers?include_state=true`

Or something similar to get the `state` field returned if you need it.

> (but I guess it should be the same setting for /users and /servers?).

Yeah this is my main question and if so, it means we should only include `state` for active servers which is how the `GET /users` (`user_model`) code handles that. I'm fine either way, was just looking for direction from maintainers on this wrinkle.

File: docs/rest-api.yml (hunk 1 above)
Comment (betatim) [reply docs/rest-api.yml]: Yes please to updating the changelog. We often forget and then someone has to go through all PRs when they are making the release.

If you look around the code some of the methods/functions where behaviour changed have a note saying when&what changed. It seems like none of the entries in `rest-api.yml` have those so I'd also not add anything for this new route. I think additions (or changes that don't break existing behaviour) are Ok to just announce in the changelog.

File: jupyterhub/tests/test_api.py [hunk 7]
```diff
@@ -176,7 +230,7 @@ def fill_user(model):
     users = [normalize_user(u) for u in users]
     assert users == [
         fill_user({'name': 'admin', 'admin': True}),
-        fill_user({'name': 'user', 'admin': False, 'last_activity': None}),
```
Comment (betatim) [inline jupyterhub/tests/test_api.py]: Why this change?

File: jupyterhub/tests/test_api.py [hunk 8]
```diff
@@ -109,6 +109,50 @@
...
+    """
+    db = app.db
+    r = await api_request(app, 'servers')
+    assert r.status_code == 200
+
+    servers = sorted(r.json(), key=lambda d: d['id'])
+    servers = [normalize_server(server) for server in servers]
+    # Note that there are two servers to start because the `app` fixture
+    # creates a user named "user" and test_referer_check creates an admin user.
+    assert servers == [fill_server(server) for server in servers]
```
Comment (betatim) [inline jupyterhub/tests/test_api.py]: What exactly does this check? At first I thought it checks that we indeed got two servers back from the API, but it doesn't. Then I thought it was checking that the model/JSON fields returned from the API call were as expected, but now I don't think it does that either.

File: jupyterhub/spawner.py [hunk 9]
```diff
@@ -212,6 +213,19 @@ def name(self):
             return self.orm_spawner.name
         return ''
 
+    # pass getattr to ORM spawner
+    def __getattr__(self, attr):
```
Comment (betatim) [inline jupyterhub/spawner.py]: Maybe @minrk can comment on why so far attributes were not automatically forwarded to the ORM class.

My Python intuition (potentially outdated) makes me feel a bit uncomfortable when I see classes implementing dunder methods for convenience. It seems like a good idea at the time but down the road it causes confusion/debugging pain. Usually the only advantage is the saving of a few characters when writing the original code.

File: jupyterhub/apihandlers/base.py [hunk 10]
```diff
@@ -131,19 +132,25 @@ def write_error(self, status_code, **kwargs):
...
+            'started': isoformat(spawner.started),
             'pending': spawner.pending,
             'ready': spawner.ready,
             'state': spawner.get_state() if include_state else None,
             'url': url_path_join(spawner.user.url, spawner.name, '/'),
             'user_options': spawner.user_options,
             'progress_url': spawner._progress_url,
         }
+        if include_user_and_id:
+            server.update({'id': spawner.id, 'user': {'name': spawner.user.name}})
```
Comment (betatim) [inline jupyterhub/apihandlers/base.py]: With `DELETE /server/{id}` out of the picture now, what do we need the ID for? To delete the server I need the username and the servername. The ID doesn't seem like something I'd need as a consumer?

If we do remove it we can rename this from the long name of `include_user_and_id` to `include_user`.

**Comment by minrk** [comment]
Would the goal here (making the request to list running users and servers more efficient for large numbers of users) perhaps be more easily accomplished with a `?active=true` query param on `GET /users` that results in a filtered query? That would be a minimal, backward-compatible API change with no new API endpoints or schema changes, just an improved efficiency for a common costly request?

File: jupyterhub/apihandlers/base.py (hunk 10 above)
Comment (mriedem) [reply jupyterhub/apihandlers/base.py]: Yeah `id` can be removed.

File: jupyterhub/spawner.py (hunk 9 above)
Comment (mriedem) [reply jupyterhub/spawner.py]: I copied this pattern from the `user.User` ORM wrapper and while it makes the client-side code for this object a bit cleaner (and less tightly coupled to the internals of this object) it does introduce some magic. I can see pros/cons both ways. I don't have an issue with removing this.

File: jupyterhub/tests/test_api.py (hunk 8 above)
Comment (mriedem) [reply jupyterhub/tests/test_api.py]: I think at some point while working on this I had it checking two servers individually like how `test_get_users` works but obviously removed that, I'm not sure why now. It should be changed since if we don't get any servers back this will still pass with a false positive. If we remove the `id` field and sort by something like the user name then we could assert two servers in the order we expect.

File: jupyterhub/tests/test_api.py (hunk 7 above)
Comment (mriedem) [reply jupyterhub/tests/test_api.py]: When I added the test above the `last_activity` field on this `user` server was no longer None, I'm not entirely sure why. I assumed it was because as a result of the new test that runs before this (and these tests aren't idempotent) that something _touched_ the server so it had an activity set, but I'm not familiar enough with the code to know what might cause that side effect.

**Comment by mriedem** [comment]
> Would the goal here (making the request to list running users and servers more efficient for large numbers of users) perhaps be more easily accomplished with a `?active=true` query param on `GET /users` that results in a filtered query? That would be a minimal, backward-compatible API change with no new API endpoints or schema changes, just an improved efficiency for a common costly request?

I replied in more detail to the same question on the issue:

https://github.com/jupyterhub/jupyterhub/issues/2954#issuecomment-596694346

tl;dr we can probably just abandon this pull request now unless you think there is other general usefulness in having this API (though if you're like me I wouldn't want to add something I need to maintain if there is no clear use case for it yet).

**Comment by mriedem** [comment]
As explained earlier and in detail in issue #2954 this is not something we're going to use now (I'll also close pull #2969) since we're now running the hub-managed `cull-idle` service with `--cull-users`. We ca follow up in issue #2954 with adding any server-side filtering logic to the `GET /users` API if we want to try and speed that up for large user data sets.
